### ISSUE TITLE
Placeholders is not properly set

### ISSUE THREAD

**Issue by vovkab** [bumptech/glide#384]
**Glide**: 3.5.2
**Device/Android Version**: any
**Screenshot**: https://www.dropbox.com/s/qexjjuv5upgl62w/Screenshot%202015-03-30%2014.33.58.png?dl=0

**Issue details/Repro steps**:
I have a grid of images, sometimes when you open it it will set place holder only to first item, evrything else will be just white square. And later it will set real images.

I've been looking through the code and looks like issue is here:

``` java
// com.bumptech.glide.request.GenericRequest#onResourceReady(Resource<?>)
public void onResourceReady(Resource<?> resource) {
    ...
    if (!canSetResource()) {
        releaseResource(resource);
        // We can't set the status to complete before asking canSetResource().
        status = Status.COMPLETE;
        return;
    }

    onResourceReady(resource, (R) received);
}

private void onResourceReady(Resource<?> resource, R result) {
    ...
    status = Status.COMPLETE;
    this.resource = resource;
    notifyLoadSuccess();
    ...
}
```

As you can see you first check `!canSetResource()` but then immediately set `status = Status.COMPLETE;` in next `onResourceReady`

**Comment by TWiStErRob** [comment]
There's a `return` there you might have missed. The way I see this: status will be COMPLETE regardless of canSetResource the question canSetResource decides if we accept it and hold onto it, or decline and drop it.

Though what you describe may actually be an issue.

**Comment by vovkab** [comment]
Yes, return is not important here. Since status will be set to complete anyway.

**Comment by vovkab** [comment]
I actually have 2 bugs:
1. placeholders is not working correctly
2. if I reopen screen, sometimes it shows placeholders even if I just loaded all images, I need to investigate this one, probably will create a new ticket, when I know more

**Comment by sjudd** [comment]
I'm not totally sure I see the issue and I'm also not able to reproduce placeholders failing to load.

What specifically do you think is wrong with setting status = COMPLETE if canSetResource returns false there? Is there any set of steps you can take that reliably reproduces the problem?

**Comment by vovkab** [comment]
@sjudd it is not easy to reproduce, but I see it a lot when I start using app. Maybe it is not issue with status complete, I just noticed that you guys have comment in code which looks like useless, since complete is always set, so you can do something like:

```
public void onResourceReady(Resource<?> resource) {
    ...
    status = Status.COMPLETE;
    if (!canSetResource()) {
        releaseResource(resource);
        return;
    }

    onResourceReady(resource, (R) received);
}
```

And it will be exactly the same what you have right now.

Now looking with debugger, I see that on first request we set placeholder, on all next since image is already in memory cache we skip this step, this is possible that this is timing issue, i.e. android start rendering images and there is nothing there, and only after you set real image. As a result we will get what you can see on my screenshot.

For me steps to reproduce is:
1. make sure new activity have a grid with all images in memory
2. start doing back and open grid again
3. repeat until you see this bug

make sure you use real device, it is much easier to reproduce on real device, then using genymotion or something else.

**Comment by sjudd** [comment]
Thanks, I'll look in to it. If you can reproduce it and happen to find a fix a pull request would definitely be welcome. I think I've seen this elsewhere once or twice too.

**Comment by sjudd** [comment]
I'm having some trouble reproducing this. Any chance you can attach a small sample app that reproduces the issue? 

Either way, can you provide the Glide load line you're using?

**Comment by nuoyan2** [comment]
身为中国人的我....我英语很差怎么办.....悲剧...给个个eclipse的工程吗

**Comment by nuoyan2** [comment]
I use asynchttp 会OOM   I use Glide Download the GIF image is very slow very slow...

**Comment by nuoyan2** [comment]
qq519719668 Please give advice or comments

**Comment by TWiStErRob** [comment]
[Next time please open a new issue. Anyway, try adding `.diskCacheStrategy(SOURCE)`.](https://translate.google.com/#en/zh-CN/Next%20time%20please%20report%20a%20new%20issue.%20Anyway%2C%20try%20to%20add%20%60.diskCacheStrategy(SOURCE)%60.)

**Comment by TWiStErRob** [comment]
@vovkab ?
